finality-aleph/src/validator_network/service.rs

```diff
@@ -150,76 +154,122 @@ impl<D: Data, A: Data + Debug, ND: Dialer<A>, NL: Listener> Service<D, A, ND, NL
-                // pass the tuple to the manager to register the connection
-                // the manager will be responsible for killing the worker if necessary
-                Some((peer_id, Some(exit), _)) = incoming_workers.next() => {
-                    use AddResult::*;
-                    match self.manager.add_incoming(peer_id.clone(), exit) {
-                        Uninterested => info!(target: "validator-network", "Peer {} connected to us despite out lack of interest.", peer_id),
-                        Added => info!(target: "validator-network", "New incoming connection for peer {}.", peer_id),
-                        Replaced => info!(target: "validator-network", "Replaced incoming connection for peer {}.", peer_id),
-                    }
-                },
-                // received information from a spawned worker managing an outgoing connection
+                // received information from a spawned worker managing a connection
                 // check if we still want to be connected to the peer, and if so, spawn a new worker or actually add proper connection
-                Some((peer_id, maybe_data_for_network, _)) = outgoing_workers.next() => {
-                    use AddResult::*;
-                    if let Some(addresses) = self.manager.peer_addresses(&peer_id) {
-                        match maybe_data_for_network {
-                            Some(data_for_network) => match self.manager.add_outgoing(peer_id.clone(), data_for_network) {
-                                Uninterested => warn!(target: "validator-network", "We connected to peer {} for unknown reasons.", peer_id),
-                                Added => info!(target: "validator-network", "New outgoing connection to peer {}.", peer_id),
-                                Replaced => info!(target: "validator-network", "Replaced outgoing connection to peer {}.", peer_id),
-                            },
-                            None => self.spawn_new_outgoing(peer_id, addresses, outgoing_result_for_parent.clone()),
+                Some((peer_id, maybe_data_for_network, connection_type)) = worker_results.next() => {
+                    use ConnectionType::*;
+                    let spawn_new_legacy_connection = match connection_type {
+                        LegacyIncoming => {
+                            self.manager.remove_peer(&peer_id);
+                            self.legacy_connected.insert(peer_id.clone())
+                        },
+                        LegacyOutgoing => {
+                            self.manager.remove_peer(&peer_id);
+                            self.legacy_connected.insert(peer_id.clone());
+                            false
                         }
+                        New => {
+                            // We always return New if connecting fails, so only New+Some means we
+                            // actually negotiated the new protocol, otherwise we should stay
```

Comment:
do we expect some tests for service?

Reply:
There are the validator network integration tests in `testing`, although they do not check compatibility. Checking compatibility would be quite a lot of work and only for temporary gain, as the compat code should be removed soon (after the next update). It's probably better to just check compatibility manually before the update (which we should do anyway).